// urban-layers-mobile/App.tsx
import { GLView } from "expo-gl";
import { useFonts } from "expo-font";
import BottomSheet, { BottomSheetFlatList, BottomSheetView } from "@gorhom/bottom-sheet";
import { DMMono_400Regular } from "@expo-google-fonts/dm-mono";
import { PlayfairDisplay_700Bold } from "@expo-google-fonts/playfair-display";
import { useCallback, useEffect, useMemo, useRef, useState } from "react";
import {
  Image,
  LayoutChangeEvent,
  Pressable,
  ScrollView,
  StyleSheet,
  Text,
  View
} from "react-native";
import MapView, { MapPressEvent, Region } from "react-native-maps";
import Slider from "@react-native-community/slider";
import * as THREE from "three";
import { GestureHandlerRootView } from "react-native-gesture-handler";
import { ReactNativeAdapter } from "./src/adapters/ReactNativeAdapter";
import { fetchRecordsByBBox } from "@urbanlayers/supabase-config";
// @ts-ignore Shared browser module reused directly.
import { resolveHeights } from "../buildingHeightResolver.js";
// @ts-ignore Shared browser module reused directly.
import { eraIndex, createMaterials } from "../material-factory.js";

type UrbanRecord = {
  id: string;
  title: string;
  category: string;
  desc: string;
  lat: number;
  lng: number;
  year: number;
  architect: string | null;
  demolished: number | null;
  img: string | null;
  height_m?: number | null;
};

type SceneState = {
  scene: THREE.Scene;
  camera: THREE.OrthographicCamera;
  meshesById: Map<string, THREE.Mesh>;
  recordByUuid: Map<string, UrbanRecord>;
  bounds: Bounds;
};

type Bounds = {
  north: number;
  south: number;
  east: number;
  west: number;
};

const METERS_TO_WORLD_UNITS = 0.00157; // mirrored from browser overlay
const INITIAL_REGION: Region = {
  latitude: 37.7749,
  longitude: -122.4194,
  latitudeDelta: 0.16,
  longitudeDelta: 0.16,
};
const SF_BOUNDS: Bounds = {
  north: 37.812,
  south: 37.703,
  east: -122.356,
  west: -122.527,
};
const DEFAULT_FETCH_RANGE = { from: 1700, to: 2100 };
const MIN_YEAR = 1791;
const MAX_YEAR = 2025;
const CATEGORIES = {
  civic: "#5b9bd5",
  commercial: "#7ec88a",
  residential: "#e8a44a",
  infrastructure: "#a782c3",
  entertainment: "#e07a7a",
  religious: "#6ecfcf",
} as const;
const ERA_PILLS = [
  { max: 1889, label: "Pre-1890" },
  { max: 1920, label: "1890-1920" },
  { max: 1945, label: "1920-1945" },
  { max: 1965, label: "1945-1965" },
  { max: Infinity, label: "Post-1965" },
];

export default function App() {
  const [fontsLoaded] = useFonts({
    PlayfairDisplay_700Bold,
    DMMono_400Regular,
  });
  const adapterRef = useRef<ReactNativeAdapter | null>(null);
  const sheetRef = useRef<BottomSheet>(null);
  const sceneRef = useRef<SceneState | null>(null);
  const [region, setRegion] = useState<Region>(INITIAL_REGION);
  const [records, setRecords] = useState<UrbanRecord[]>([]);
  const [viewYear, setViewYear] = useState(1950);
  const [activeCategories, setActiveCategories] = useState(new Set(Object.keys(CATEGORIES)));
  const [selectedRecord, setSelectedRecord] = useState<UrbanRecord | null>(null);
  const [viewport, setViewport] = useState({ width: 1, height: 1 });

  const mapInitialRegion = useMemo(() => INITIAL_REGION, []);
  const snapPoints = useMemo(() => ["80", "58%", "100%"], []);
  const currentEra = useMemo(
    () => ERA_PILLS.find((entry) => viewYear <= entry.max)?.label ?? "Post-1965",
    [viewYear]
  );
  const visibleRecords = useMemo(
    () => records.filter((record) => {
      if (!activeCategories.has(record.category)) return false;
      if (record.year > viewYear) return false;
      if (record.demolished !== null && record.demolished <= viewYear) return false;
      return true;
    }),
    [activeCategories, records, viewYear]
  );

  const syncCameraToRegion = useCallback((nextRegion: Region) => {
    const sceneState = sceneRef.current;
    if (!sceneState) return;

    const bounds = regionToBounds(nextRegion);
    sceneState.bounds = bounds;

    const aspect = Math.max(viewport.width / Math.max(viewport.height, 1), 1);
    sceneState.camera.left = -aspect;
    sceneState.camera.right = aspect;
    sceneState.camera.top = 1;
    sceneState.camera.bottom = -1;
    sceneState.camera.position.set(0, -0.4, 0.915);
    sceneState.camera.lookAt(0, 0, 0);
    sceneState.camera.updateProjectionMatrix();

    for (const [id, mesh] of sceneState.meshesById.entries()) {
      const record = sceneState.recordByUuid.get(id);
      if (!record) continue;
      const projected = projectToWorld(record, bounds, aspect);
      mesh.position.x = projected.x;
      mesh.position.y = projected.y;
    }
  }, [records, viewport.height, viewport.width]);

  const buildSceneMeshes = useCallback(async (sceneState: SceneState, recs: UrbanRecord[]) => {
    for (const mesh of sceneState.meshesById.values()) {
      sceneState.scene.remove(mesh);
      mesh.geometry.dispose();
    }
    sceneState.meshesById.clear();
    sceneState.recordByUuid.clear();

    const materialSet = createMaterials(THREE);
    let heights = new Map<string, number>();
    try {
      heights = await resolveHeights(recs);
    } catch (error) {
      console.warn("[UrbanLayersMobile] Height resolution failed:", error);
    }

    const aspect = Math.max(viewport.width / Math.max(viewport.height, 1), 1);
    for (const record of recs) {
      const era = eraIndex(record.year);
      const heightM = heights.get(record.id) ?? record.height_m ?? 10;
      const height = Math.max(heightM * METERS_TO_WORLD_UNITS, 0.015);
      const geometry = new THREE.BoxGeometry(0.03, 0.03, height);
      const mesh = new THREE.Mesh(geometry, materialSet[era].solid);
      const projected = projectToWorld(record, sceneState.bounds, aspect);
      mesh.position.set(projected.x, projected.y, height / 2);
      sceneState.scene.add(mesh);
      sceneState.meshesById.set(mesh.uuid, mesh);
      sceneState.recordByUuid.set(mesh.uuid, record);
    }
  }, [viewport.height, viewport.width]);

  const handleContextCreate = useCallback(async (gl: WebGLRenderingContext) => {
    const adapter = new ReactNativeAdapter(gl);
    adapterRef.current = adapter;
    await adapter.initialize();

    const scene = new THREE.Scene();
    scene.background = null;

    const camera = new THREE.OrthographicCamera(-adapter.aspect, adapter.aspect, 1, -1, -20, 20);
    camera.position.set(0, -0.4, 0.915);
    camera.lookAt(0, 0, 0);

    scene.add(new THREE.AmbientLight(0xffe8d0, 0.55));
    const keyLight = new THREE.DirectionalLight(0xffb266, 1.1);
    keyLight.position.set(-0.6, -0.8, 1.5);
    scene.add(keyLight);

    sceneRef.current = {
      scene,
      camera,
      meshesById: new Map<string, THREE.Mesh>(),
      recordByUuid: new Map<string, UrbanRecord>(),
      bounds: regionToBounds(region),
    };

    if (visibleRecords.length > 0) {
      await buildSceneMeshes(sceneRef.current, visibleRecords);
      syncCameraToRegion(region);
    }

    adapter.start(() => {
      adapter.renderFrame(scene, camera);
    });
  }, [buildSceneMeshes, region, syncCameraToRegion, visibleRecords]);

  const handleRegionChange = useCallback((nextRegion: Region) => {
    setRegion(nextRegion);
    syncCameraToRegion(nextRegion);
  }, [syncCameraToRegion]);

  const handleMapPress = useCallback((event: MapPressEvent) => {
    const sceneState = sceneRef.current;
    if (!sceneState) return;
    const hit = raycastFromMapPress(
      event.nativeEvent.coordinate.latitude,
      event.nativeEvent.coordinate.longitude,
      sceneState.bounds,
      sceneState.camera,
      Array.from(sceneState.meshesById.values()),
      viewport.width / Math.max(viewport.height, 1)
    );
    if (!hit) return;
    const tapped = sceneState.recordByUuid.get(hit.object.uuid) ?? null;
    if (!tapped) return;
    setSelectedRecord(tapped);
    sheetRef.current?.snapToIndex(2);
  }, [viewport.height, viewport.width]);

  const handleLayout = useCallback((event: LayoutChangeEvent) => {
    const width = Math.max(event.nativeEvent.layout.width, 1);
    const height = Math.max(event.nativeEvent.layout.height, 1);
    setViewport({ width, height });
    adapterRef.current?.resize(width, height);
    syncCameraToRegion(region);
  }, [region, syncCameraToRegion]);

  const fetchRecords = useCallback(async () => {
    try {
      const fetched = await fetchRecordsByBBox(SF_BOUNDS, DEFAULT_FETCH_RANGE);
      setRecords(fetched as UrbanRecord[]);
    } catch (error) {
      console.warn("[UrbanLayersMobile] Supabase fetch failed:", error);
      setRecords([]);
    }
  }, []);

  useEffect(() => {
    fetchRecords();
  }, [fetchRecords]);

  useEffect(() => {
    if (!sceneRef.current) return;
    buildSceneMeshes(sceneRef.current, visibleRecords).then(() => {
      syncCameraToRegion(region);
    });
  }, [buildSceneMeshes, region, syncCameraToRegion, visibleRecords]);

  useEffect(() => {
    return () => {
      if (sceneRef.current) {
        for (const mesh of sceneRef.current.meshesById.values()) {
          mesh.geometry.dispose();
        }
        sceneRef.current.meshesById.clear();
      }
      adapterRef.current?.dispose();
      adapterRef.current = null;
    };
  }, []);

  const toggleCategory = useCallback((category: string) => {
    setActiveCategories((prev) => {
      const next = new Set(prev);
      if (next.has(category)) {
        next.delete(category);
      } else {
        next.add(category);
      }
      return next;
    });
  }, []);

  const onSheetChange = useCallback((index: number) => {
    if (index !== 2 && selectedRecord) {
      setSelectedRecord(null);
    }
  }, [selectedRecord]);

  if (!fontsLoaded) {
    return <View style={styles.container} />;
  }

  return (
    <GestureHandlerRootView style={styles.container}>
      <View style={styles.container} onLayout={handleLayout}>
        <MapView
          style={StyleSheet.absoluteFillObject}
          initialRegion={mapInitialRegion}
          onRegionChange={handleRegionChange}
          onPress={handleMapPress}
        />
        <GLView
          style={styles.glOverlay}
          pointerEvents="none"
          onContextCreate={handleContextCreate}
        />
        <BottomSheet
          ref={sheetRef}
          index={0}
          snapPoints={snapPoints}
          onChange={onSheetChange}
          backgroundStyle={styles.sheetBackground}
          handleIndicatorStyle={styles.sheetHandle}
        >
          {selectedRecord ? (
            <BottomSheetView style={styles.detailWrapper}>
              <ScrollView contentContainerStyle={styles.detailScrollContent}>
                <Text style={styles.detailTitle}>{selectedRecord.title}</Text>
                <Text style={styles.detailMeta}>
                  {selectedRecord.year} · {selectedRecord.category}
                </Text>
                <Text style={styles.detailBody}>{selectedRecord.desc || "No description available."}</Text>
                <Text style={styles.detailBody}>Architect: {selectedRecord.architect ?? "Unknown"}</Text>
                <Text style={styles.detailBody}>
                  Demolished: {selectedRecord.demolished ?? "Still standing"}
                </Text>
                {selectedRecord.img ? (
                  <Image source={{ uri: selectedRecord.img }} style={styles.detailImage} />
                ) : (
                  <View style={styles.detailImageFallback}>
                    <Text style={styles.detailFallbackText}>No historical photo</Text>
                  </View>
                )}
                <Pressable
                  style={styles.dismissButton}
                  onPress={() => {
                    setSelectedRecord(null);
                    sheetRef.current?.snapToIndex(1);
                  }}
                >
                  <Text style={styles.dismissButtonText}>Dismiss</Text>
                </Pressable>
              </ScrollView>
            </BottomSheetView>
          ) : (
            <BottomSheetView style={styles.sheetContent}>
              <View style={styles.peekRow}>
                <Text style={styles.peekYear}>{viewYear}</Text>
                <Text style={styles.peekEra}>{currentEra}</Text>
                <Text style={styles.peekCount}>{visibleRecords.length} records</Text>
              </View>
              <View style={styles.expandedControls}>
                <Text style={styles.sectionLabel}>Year</Text>
                <Slider
                  minimumValue={MIN_YEAR}
                  maximumValue={MAX_YEAR}
                  step={1}
                  value={viewYear}
                  minimumTrackTintColor="#d4a857"
                  maximumTrackTintColor="#3a3020"
                  thumbTintColor="#d4a857"
                  onValueChange={(value) => setViewYear(value)}
                />
                <View style={styles.categoryRow}>
                  {Object.entries(CATEGORIES).map(([category, color]) => {
                    const active = activeCategories.has(category);
                    return (
                      <Pressable
                        key={category}
                        style={[
                          styles.categoryChip,
                          { borderColor: color, backgroundColor: active ? `${color}33` : "transparent", opacity: active ? 1 : 0.45 }
                        ]}
                        onPress={() => toggleCategory(category)}
                      >
                        <Text style={styles.categoryChipText}>{category}</Text>
                      </Pressable>
                    );
                  })}
                </View>
                <Text style={styles.sectionLabel}>Records</Text>
                <BottomSheetFlatList
                  data={visibleRecords}
                  keyExtractor={(item) => item.id}
                  style={styles.recordList}
                  renderItem={({ item }) => (
                    <Pressable
                      onPress={() => {
                        setSelectedRecord(item);
                        sheetRef.current?.snapToIndex(2);
                      }}
                      style={styles.recordRow}
                    >
                      <Text style={styles.recordTitle}>{item.title}</Text>
                      <Text style={styles.recordMeta}>{item.year} · {item.category}</Text>
                    </Pressable>
                  )}
                />
              </View>
            </BottomSheetView>
          )}
        </BottomSheet>
      </View>
    </GestureHandlerRootView>
  );
}

function regionToBounds(nextRegion: Region): Bounds {
  const halfLat = nextRegion.latitudeDelta / 2;
  const halfLng = nextRegion.longitudeDelta / 2;
  return {
    north: nextRegion.latitude + halfLat,
    south: nextRegion.latitude - halfLat,
    east: nextRegion.longitude + halfLng,
    west: nextRegion.longitude - halfLng,
  };
}

function projectToWorld(record: Pick<UrbanRecord, "lat" | "lng">, bounds: Bounds, aspect: number) {
  const xNorm = (record.lng - bounds.west) / Math.max(bounds.east - bounds.west, 1e-9);
  const yNorm = (record.lat - bounds.south) / Math.max(bounds.north - bounds.south, 1e-9);
  return {
    x: ((xNorm * 2) - 1) * aspect,
    y: -(((1 - yNorm) * 2) - 1),
  };
}

function raycastFromMapPress(
  lat: number,
  lng: number,
  bounds: Bounds,
  camera: THREE.Camera,
  meshes: THREE.Mesh[],
  aspect: number
) {
  if (meshes.length === 0) return null;
  const projected = projectToWorld({ lat, lng }, bounds, aspect);
  const ndc = new THREE.Vector2(projected.x / aspect, -projected.y);
  const raycaster = new THREE.Raycaster();
  raycaster.setFromCamera(ndc, camera);
  const hits = raycaster.intersectObjects(meshes);
  return hits[0] ?? null;
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#0e0e1a",
  },
  glOverlay: {
    ...StyleSheet.absoluteFillObject,
  },
  sheetBackground: {
    backgroundColor: "rgba(8,8,18,0.97)",
    borderTopColor: "#3a3020",
    borderTopWidth: 1,
  },
  sheetHandle: {
    backgroundColor: "#d4a857",
  },
  sheetContent: {
    flex: 1,
    paddingHorizontal: 16,
    paddingBottom: 16,
  },
  peekRow: {
    minHeight: 80,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
  },
  peekYear: {
    color: "#d4a857",
    fontFamily: "DMMono_400Regular",
    fontSize: 26,
  },
  peekEra: {
    color: "#e0d5c5",
    fontFamily: "PlayfairDisplay_700Bold",
    fontSize: 16,
  },
  peekCount: {
    color: "#888888",
    fontFamily: "DMMono_400Regular",
    fontSize: 14,
  },
  expandedControls: {
    flex: 1,
    gap: 10,
  },
  sectionLabel: {
    color: "#d4a857",
    fontFamily: "DMMono_400Regular",
    textTransform: "uppercase",
    letterSpacing: 1.2,
    fontSize: 12,
  },
  categoryRow: {
    flexDirection: "row",
    flexWrap: "wrap",
    gap: 8,
  },
  categoryChip: {
    borderWidth: 1,
    borderRadius: 14,
    paddingHorizontal: 10,
    paddingVertical: 5,
  },
  categoryChipText: {
    color: "#e0d5c5",
    fontFamily: "DMMono_400Regular",
    fontSize: 11,
    textTransform: "capitalize",
  },
  recordList: {
    flex: 1,
  },
  recordRow: {
    borderBottomColor: "#3a3020",
    borderBottomWidth: 1,
    paddingVertical: 10,
  },
  recordTitle: {
    color: "#e0d5c5",
    fontFamily: "PlayfairDisplay_700Bold",
    fontSize: 16,
  },
  recordMeta: {
    color: "#888888",
    fontFamily: "DMMono_400Regular",
    fontSize: 12,
  },
  detailWrapper: {
    flex: 1,
  },
  detailScrollContent: {
    paddingHorizontal: 18,
    paddingBottom: 40,
    gap: 10,
  },
  detailTitle: {
    color: "#d4a857",
    fontFamily: "PlayfairDisplay_700Bold",
    fontSize: 30,
  },
  detailMeta: {
    color: "#888888",
    fontFamily: "DMMono_400Regular",
    fontSize: 13,
  },
  detailBody: {
    color: "#e0d5c5",
    fontFamily: "DMMono_400Regular",
    fontSize: 14,
    lineHeight: 21,
  },
  detailImage: {
    width: "100%",
    height: 220,
    borderRadius: 8,
    borderColor: "#3a3020",
    borderWidth: 1,
  },
  detailImageFallback: {
    width: "100%",
    height: 220,
    borderRadius: 8,
    borderColor: "#3a3020",
    borderWidth: 1,
    backgroundColor: "#16162a",
    alignItems: "center",
    justifyContent: "center",
  },
  detailFallbackText: {
    color: "#666666",
    fontFamily: "DMMono_400Regular",
  },
  dismissButton: {
    borderColor: "#d4a857",
    borderWidth: 1,
    borderRadius: 7,
    paddingVertical: 10,
    alignItems: "center",
    marginTop: 10,
  },
  dismissButtonText: {
    color: "#d4a857",
    fontFamily: "DMMono_400Regular",
    fontSize: 13,
    letterSpacing: 1.2,
  },
});
